hazelcast-client/src/main/java/com/hazelcast/client/config/ClientFailoverClasspathYamlConfig.java

```diff
@@ -0,0 +1,88 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.client.config;
+
+import com.hazelcast.core.HazelcastException;
+import com.hazelcast.logging.ILogger;
+import com.hazelcast.logging.Logger;
+
+import java.io.InputStream;
+import java.util.Properties;
+
+/**
+ * A {@link ClientFailoverConfig} which is initialized by loading a YAML
+ * configuration file from the classpath.
+ */
+public class ClientFailoverClasspathYamlConfig extends ClientFailoverConfig {
+    private static final ILogger LOGGER = Logger.getLogger(ClientFailoverClasspathYamlConfig.class);
+
+    /**
+     * Creates a config which is loaded from a classpath resource using the
+     * Thread.currentThread contextClassLoader. The System.properties are used to resolve variables
+     * in the YAML.
+     *
+     * @param resource the resource, an YAML configuration file from the classpath
+     * @throws IllegalArgumentException if the resource could not be found
+     * @throws HazelcastException       if the YAML content is invalid
+     */
+    public ClientFailoverClasspathYamlConfig(String resource) {
+        this(resource, System.getProperties());
+    }
+
+    /**
+     * Creates a config which is loaded from a classpath resource using the
+     * Thread.currentThread contextClassLoader.
+     *
+     * @param resource   the resource, an YAML configuration file from the classpath
+     * @param properties the Properties to resolve variables in the YAML
+     * @throws IllegalArgumentException if the resource could not be found or if properties is {@code null}
+     * @throws HazelcastException       if the YAML content is invalid
+     */
+    public ClientFailoverClasspathYamlConfig(String resource, Properties properties) {
+        this(Thread.currentThread().getContextClassLoader(), resource, properties);
+    }
+
+    /**
+     * Creates a config which is loaded from a classpath resource.
+     *
+     * @param classLoader the ClassLoader used to load the resource
+     * @param resource    the resource, an YAML configuration file from the classpath
+     * @param properties  the properties used to resolve variables in the YAML
+     * @throws IllegalArgumentException if classLoader or resource is {@code null}, or if the resource is not found
+     * @throws HazelcastException       if the YAML content is invalid
+     */
+    public ClientFailoverClasspathYamlConfig(ClassLoader classLoader, String resource, Properties properties) {
+        if (classLoader == null) {
```

Comment:
Unfortunately we can't, because `checkNotNull` throws NPE in this case, while our contract is `IllegalArgumentException` for these cases.

Reply:
How did you decide on the contract, because I see some configs used `Preconditions.checkNotNull`(e.g.  in the code base CardinalityEstimatorConfig.setName, UsernamePasswordCredentials constructor, ReliableTopicConfig constructor, etc.) ?

Reply:
It is the contract of all config classes, and I think we should keep failover configs to be consistent with other client and member configs. 

I agree that these inconsistencies are not good, but addressing them is outside of the scope of this PR. Created an issue to address this inconsistency of the Config API: https://github.com/hazelcast/hazelcast/issues/14713